src/Streamly/Internal/Data/Zipper.hs

```diff
@@ -0,0 +1,380 @@
+-- data in) a higher level zipper could be useful in implementing distributive
+-- parsers.
+--
+-- See https://hackage.haskell.org/package/zipper for an existing type class.
+-- We need a monadic one.
+
+module Streamly.Internal.Data.Zipper
+    (
+      Zipper (..)
+
+    -- * Construction
+    , nil
+    , fromStream
+    , fromList
+
+    -- * Checkpointing
+    , checkpoint
+    , release
+    , restore
+
+    -- * Parsing
+    , parse
+    )
+where
+
+import Control.Exception (assert, Exception(..))
+import Control.Monad.Catch (MonadCatch, try, SomeException)
+-- import GHC.Types (SPEC(..))
+import Prelude hiding (splitAt)
+
+import Streamly.Internal.Data.Stream.StreamK.Type (Stream(..))
+import Streamly.Internal.Data.SVar (defState)
+
+import qualified Streamly.Internal.Data.Parser.ParserD.Types as PR
+import qualified Streamly.Internal.Data.Stream.StreamK as K
+
+-------------------------------------------------------------------------------
+-- Zipper type
+-------------------------------------------------------------------------------
+
+-- | @Zipper checkpoints lefts rights tail@.  The focus is on the first element
+-- of @rights@.  @lefts@ is buffered data on the right of the cursor.  @tail@
+-- is a stream that is used to generate more data if the cursor moves past
+-- @rights@.
+--
+-- @checkpoints@ is a stack of checkpoints. A new checkpoint is created by a
+-- @checkpoint@ operation. A checkpoint consists of a count that tracks how
+-- many elements we have yielded after the checkpoint was taken.  We need this
+-- information to backtrack to the checkpoint.  If we enter a nested
+-- alternative we add another checkpoint in the stack.  When we exit an
+-- alternative we call a @release@ on the checkpoint. The @release@ removes the
+-- checkpoint from the stack and adds its element count to the previous
+-- checkpoint in the stack. When the last checkpoint is removed the buffer is
+-- released.
+--
+-- /Internal/
+--
+data Zipper m a = Zipper [Int] [a] [a] (Stream m a)
+
+-------------------------------------------------------------------------------
+-- Construction
+-------------------------------------------------------------------------------
+
+-- | Empty zipper.
+--
+-- /Internal/
+--
+{-# INLINE nil #-}
+nil :: Zipper m a
+nil = Zipper [] [] [] K.nil
+
+-- | Create a zipper from a stream.
+--
+-- /Internal/
+--
+{-# INLINE fromStream #-}
+fromStream :: Stream m a -> Zipper m a
+fromStream = Zipper [] [] []
+
+-- | Create a zipper from a list.
+--
+-- /Internal/
+--
+{-# INLINE fromList #-}
+fromList :: [a] -> Zipper m a
+fromList xs = Zipper [] [] xs K.nil
+
+-- | Add a checkpoint to the Zipper so that we do not release the buffer beyond
+-- the checkpoint.
+--
+-- /Internal/
+--
+{-# INLINE checkpoint #-}
+checkpoint :: Zipper m a -> Zipper m a
+checkpoint (Zipper cps xs ys stream) = Zipper (0:cps) xs ys stream
+
+-- | Release the latest checkpoint, releases any values held by the checkpoint.
```

Comment:
It does not necessarily release values held by the checkpoint. Or did I misunderstand what releasing values mean? According to me, releasing values = discarding values from the buffer.

Reply:
It releases the latest checkpoint, if the values are held by that checkpoint they are released from that checkpoint's perspective but the values may be held by other checkpoints so they will keep holding those. I am adding this note:

```
-- Note that the values may still be held by other checkpoints in the stack of
-- checkpoints.
```